src/Features/Core/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs

```diff
@@ -22,111 +26,200 @@ public DiagnosticIncrementalAnalyzer(DiagnosticAnalyzerService owner, int correl
         }
 
         #region IIncrementalAnalyzer
-        public override Task AnalyzeDocumentAsync(Document document, SyntaxNode bodyOpt, CancellationToken cancellationToken)
+        public async override Task AnalyzeDocumentAsync(Document document, SyntaxNode body, CancellationToken cancellationToken)
         {
+            // ToDo: Should there be an exception handler here? Should there be checks for cancellation?
+
+            if (document.SupportsSemanticModel)
+            {
+                Project project = document.Project;
+                VersionStamp projectVersion = await project.GetDependentVersionAsync(cancellationToken).ConfigureAwait(false);
+
+                Compilation compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                CompilationResult compilationResult = GetCompilationResult(compilation, project, projectVersion, cancellationToken);
+
+                if (!compilationResult.DocumentAnalyzed(document.Id) && !compilationResult.CompletionStarted)
+                {
+                    ImmutableArray<Diagnostic> diagnostics = await compilationResult.AnalyzeDocumentAsync(document, cancellationToken).ConfigureAwait(false);
+                    RaiseEvents(project, document, diagnostics);
+                }
+            }
+        }
+
+        public override Task AnalyzeSyntaxAsync(Document document, CancellationToken cancellationToken)
+        {
+            // Nothing happens here. The syntax tree analyzers get run as part of AnalyzeDocumentAsync.
             return SpecializedTasks.EmptyTask;
```

Comment:
Yes, if the compiler logic for syntax analysis can be teased into a separate task.